db/src/types.rs

```diff
@@ -0,0 +1,251 @@
+/// such a positive integer an *entid*.
+///
+/// ## Partitions
+///
+/// Datomic partitions the entid space in order to separate core knowledge base entities required
+/// for the healthy function of the system from user-defined entities.  Datomic also partitions in
+/// order to ensure that certain index walks of related entities are efficient.  Mentat follows
+/// suit, partitioning into the following partitions:
+/// * `:db.part/db`, for core knowledge base entities;
+/// * `:db.part/user`, for user-defined entities;
+/// * `:db.part/tx`, for transaction entities.
+/// You almost certainly want to add new entities in the `:db.part/user` partition.
+///
+/// The entid sequence in a given partition is monotonically increasing, although not necessarily
+/// contiguous.  That is, it is possible for a specific entid to have never been present in the
+/// system, even though its predecessor and successor is present.
+
+// #[derive(Debug)]
+// pub enum Error {
+//     RusqliteError(rusqlite::Error),
+//     BadSchemaAssertion(String),
+//     BadBootstrapDefinition(String),
+// }
+
+// impl From<rusqlite::Error> for Error {
+//     fn from(e: rusqlite::Error) -> Error {
+//         Error::RusqliteError(e)
+//     }
+// }
+
+/// Represents one entid in the entid space.
+///
+/// Per https://www.sqlite.org/datatype3.html (see also http://stackoverflow.com/a/8499544), SQLite
+/// stores signed integers up to 64 bits in size.  Since u32 is not appropriate for our use case, we
+/// use i64 rather than manually truncating u64 to u63 and casting to i64 throughout the codebase.
+pub type Entid = i64;
+
+/// The attribute of each Mentat assertion has an :db/valueType constraining the value to a
+/// particular domain.  Mentat recognizes the following :db/valueType values.
+#[derive(Clone,Debug,Eq,Hash,Ord,PartialOrd,PartialEq)]
+pub enum ValueType {
+    Ref,
+    Boolean,
+    Instant,
+    Long,
+    Double,
+    String,
+    UUID,
+    URI,
+    Keyword,
+}
+
+impl ValueType {
+    pub fn value_type_tag(&self) -> i32 {
+        match *self {
+            ValueType::Ref => 0,
+            ValueType::Boolean => 1,
+            ValueType::Instant => 4,
+            ValueType::Long => 5, // SQLite distinguishes integral from decimal types, allowing long and double to share a tag.
+            ValueType::Double => 5, // SQLite distinguishes integral from decimal types, allowing long and double to share a tag.
+            ValueType::String => 10,
+            ValueType::UUID => 11,
+            ValueType::URI => 12,
+            ValueType::Keyword => 13,
+        }
+    }
+
+    pub fn from_value_type_tag(value_type_tag: &i32) -> Option<ValueType> {
+        match *value_type_tag {
+            0 => Some(ValueType::Ref),
+            1 => Some(ValueType::Boolean),
+            4 => Some(ValueType::Instant),
+            5 => Some(ValueType::Long), // SQLite distinguishes integral from decimal types, allowing long and double to share a tag.
```

Comment:
You might consider avoiding this unidirectional and impossible conversion, which requires lying here and checking elsewhere.

That is: you can turn `(SQLite value, tag)` into a `ValueType`; you can turn a `ValueType` into a `tag`; you can turn a `tag` into a *set* of `ValueType`s; you cannot turn a `tag` into a single `ValueType`.

You could define a predicate, `uniquely_identifies_value_type`, if you wish.

This becomes more important when we get to querying.